fdbserver/DataDistribution.actor.cpp

```diff
@@ -1056,104 +1421,749 @@ struct DDTeamCollection {
+//						.detail("Attempt", i).detail("ServerInfoSize", server_info.size())
+//						.detail("TeamSize", team.size()).detail("UID", (**process).toString())
+//						.detail("ProcessIP", server_info.find(**process) != server_info.end()
+//								? server_info[**process]->lastKnownInterface.address().toString() : "[unset]");
 					score += server_info[**process]->teams.size();
+					Standalone<StringRef> machine_id  = server_info[**process]->lastKnownInterface.locality.machineId().get();
+					machineIDs.push_back(machine_id);
+				}
+				if ( !isMachineIDValid(machineIDs) ) {
+					maxAttempts += 1;
+					continue;
+				}
+				if ( !isMachineTeamHealthy(machineIDs) ) {
+					TraceEvent(SevError, "MachineTeamUnhealthy_ShouldNeverHappenHere")
+						.detail("Primary", primary).detail("MachineIDSize", machineIDs.size());
+					traceAllInfo(true);
+					maxAttempts += 1;
+					continue;
+				}
+				std::sort(machineIDs.begin(), machineIDs.end());
+				if ( machineTeamExists(machineIDs) ) {
+					maxAttempts += 1;
+					continue;
 				}
 
-				if(score < bestScore) {
-					bestTeam = team;
+				// SOMEDAY: randomly pick one from teams with the lowest score
+				if ( score < bestScore ) {
+					bestTeam = team; // bestTeam is the team which has the smallest number of teams its team members belong to.
 					bestScore = score;
 				}
 			}
 
-			if( bestTeam.size() == configuration.storageTeamSize) {
-				vector<UID> processIDs;
+			//Step 5: Restore machine from its representative process team and get the machine team
+			if ( bestTeam.size() == configuration.storageTeamSize ) {
+				vector<Standalone<StringRef>> machineIDs;
+				for ( auto process = bestTeam.begin(); process < bestTeam.end(); process++ ) {
+					Standalone<StringRef> machine_id  = server_info[**process]->lastKnownInterface.locality.machineId().get();
+					assert( machine_info.find(machine_id) != machine_info.end() );
+					machineIDs.push_back(machine_id);
+				}
+
+				std::sort(machineIDs.begin(), machineIDs.end());
+				if( !machineTeamExists(machineIDs) ) {
+					addMachineTeam(machineIDs.begin(), machineIDs.end());
+					addedMachineTeams++;
+				}
+			} else {
+				TraceEvent(SevWarn, "DataDistributionBuildTeams", masterId).detail("Primary", primary)
+					.detail("Reason","Unable to make desired machine Teams");
+				break;
+			}
+
+			if ( ++loopCount > 2*machineTeamsToBuild*(configuration.storageTeamSize+1) ) { // MX: Q: why is this number?
+				break;
+			}
+		}
+
+		return addedMachineTeams;
+	}
+
+	bool isMachineTeamHealthy( vector<Standalone<StringRef>> const &machineIDs ) {
+		int healthyNum = 0;
+
+		// A healthy machine team should have the desired number of machines
+		if ( machineIDs.size() != configuration.storageTeamSize )
+			return false;
+
+		for ( auto &id : machineIDs ) {
+			auto &machine = machine_info[id];
+			if ( isMachineHealthy(machine) ) {
+				healthyNum++;
+			}
+		}
+		if ( healthyNum == machineIDs.size() )
```

Comment:
`if ( expr ) { return true; } else { return false; }` should always just be `return expr;`